contracts/utils.go

```diff
@@ -4,85 +4,150 @@ import (
 	"bytes"
 	"errors"
 	"math/big"
+	"time"
 
-	abipkg "github.com/celo-org/celo-blockchain/accounts/abi"
+	"github.com/celo-org/celo-blockchain/accounts/abi"
 	"github.com/celo-org/celo-blockchain/common"
 	"github.com/celo-org/celo-blockchain/common/hexutil"
 	"github.com/celo-org/celo-blockchain/core/vm"
 	"github.com/celo-org/celo-blockchain/log"
+	"github.com/celo-org/celo-blockchain/metrics"
 )
 
-// systemCaller is the caller when the EVM is invoked from the within the blockchain system.
-var systemCaller = vm.AccountRef(common.HexToAddress("0x0"))
+// SystemCaller is the caller when the EVM is invoked from the within the blockchain system.
+var SystemCaller = common.HexToAddress("0x0")
 
-func StaticCallFromSystem(evm *vm.EVM, contractAddress common.Address, abi abipkg.ABI, funcName string, args []interface{}, returnObj interface{}, gas uint64) (uint64, error) {
-	staticCall := func(transactionData []byte) ([]byte, uint64, error) {
-		return evm.StaticCall(systemCaller, contractAddress, transactionData, gas)
-	}
+type Contract struct {
+	abi     *abi.ABI
+	address common.Address
+	caller  common.Address
+}
 
-	return handleABICall(evm, abi, funcName, args, returnObj, staticCall)
+type Backend interface {
+	Call(caller vm.ContractRef, addr common.Address, input []byte, gas uint64, value *big.Int) (ret []byte, leftOverGas uint64, err error)
+	StaticCall(caller vm.ContractRef, addr common.Address, input []byte, gas uint64) (ret []byte, leftOverGas uint64, err error)
+	StopGasMetering()
+	StartGasMetering()
+	GetStateDB() vm.StateDB
 }
 
-func CallFromSystem(evm *vm.EVM, contractAddress common.Address, abi abipkg.ABI, funcName string, args []interface{}, returnObj interface{}, gas uint64, value *big.Int) (uint64, error) {
-	call := func(transactionData []byte) ([]byte, uint64, error) {
-		return evm.Call(systemCaller, contractAddress, transactionData, gas, value)
+func NewContract(contractAbi *abi.ABI, contractAddress, defaultCaller common.Address) *Contract {
+	return &Contract{
+		abi:     contractAbi,
+		address: contractAddress,
```

Comment:
There is something a little strange about the address being here, since the address (except in the case of the registry, where it's hard-coded) is really part of the state and thus part of the backend that the contract will be used with.  But I don't see another way given that one of the uses is `MakeStaticCallWithAddress` which needs to be callable with ERC20 tokens referenced by address rather than contracts listed in the registry.